### DISCUSSION TITLE
Allow spawning of orphaned actors

### DISCUSSION THREAD

**Discussion by mdpratt** [statelyai/xstate#3179] [Ideas] [answered]
Per the [current documentation](https://xstate.js.org/docs/guides/actors.html#spawning-actors), you can only spawn actors within an `assign` action:

>Just as in Actor-model-based languages like [Akka (opens new window)](https://doc.akka.io/docs/akka/current/guide/introduction.html)or [Erlang (opens new window)](http://www.erlang.org/docs), actors are spawned and referenced in context (as the result of an assign(...) action).
>  
> Import the spawn function from 'xstate'
> 1. In an assign(...) action, create a new actor reference with spawn(...)
> 2. The spawn(...) function creates an actor reference by providing 1 or 2 arguments:

However, there are many chases where spawning an orphaned actor is the intent. For example, a notification system may wish to spawn a new notification, and then let the child handle it's own lifecycle and context.

There are some expectations with orphaned actors:
1. They would not be added to the `machine.children` set

Some ideas for handling orphaned actors:
1. Create a new `spawnOrphan` action
2. Add an `{ orphan: true}` property to the existing `spawn` action
3. Allow the `spawn` action to be called outside of the `assign` action

**Comment by Andarist** [comment]
> However, there are many chases where spawning an orphaned actor is the intent. For example, a notification system may wish to spawn a new notification, and then let the child handle it's own lifecycle and context.

Shouldn't the notifications be "hoisted" then and managed elsewhere (semi-globally)? This way you would just send an event to that "semi-global" actor and its lifetime wouldn't be bound to the machine that "triggered" the notification.

**Reply by mdpratt** [reply]
I think that depends on the implementation of the notification system. For example, if we look at something like the [AWS SNS](https://aws.amazon.com/sns/), the service acts as an orchestrator for creation only. Once the event is broadcast then it forgets the event.

**Reply by Andarist** [reply]
If it shouldn't be "tied" to the machine responsible for creating it - why can't you just use an action and `interpret` within it? Do you need the return value of `spawn`?

**Reply by EthanEFung** [reply]
I too am confused about why a machine should be responsible for spawning an actor, and yet not responsible for managing it. In my mind the orchestrator should be responsible for receiving the event to spawn, and spawning the actor.

**Reply by Andarist** [reply]
@EthanEFung could you write a pseudo-code of how would you like to use it?

**Reply by EthanEFung** [reply]
Just to be clear here, I'm in agreement with you @Andarist . To me it doesn't make sense to spawn orphaned actors. It makes practical sense that if anything were to go wrong, there would be some entity the spawned actor can notify. If this is the case then why not notify a parent actor? How would a spawned actor benefit from being orphaned?

take for example SNS - I gave it my best try trying to model the behavior
https://stately.ai/viz/fac8ea28-8027-423c-a484-9a170709e37a

Where my knowledge gap begins is 'invoking' the service. Really I believe that these two actors would be responsible handling processing events, but wasn't sure how the orchestrator would pass the messages to each actor.

**Reply by mdpratt** [reply]
Thanks, I've been meaning to post a response here. I think this is simply one of those "you can't see the tree in a forest", where trying to solve an XState "problem", when you could just solve it like a normal person.

You callouts here are valid. Oops!

**Comment by mdpratt** [comment, marked as answer]
Per the conversation above, this could easily be solved by just using `interpret` within an action. Closing this out.
